pysecs/secs.py

```diff
@@ -121,39 +121,41 @@ def fit(self, obs_loc, obs_B, obs_var=None, epsilon=0.05):
 
         # Calculate the singular value decomposition (SVD)
         # NOTE: T_obs has shape (nobs, 3, nsec), we reshape it
-        # to (nobs*3, nsec); obs_var has shape (ntimes, nobs, 3), 
+        # to (nobs*3, nsec); obs_std has shape (ntimes, nobs, 3), 
         # we reshape it to (ntimes, nobs*3), then loop over ntimes
         # to solve using (potentially) time-dependent observation 
-        # error variances to weight the observations
+        # standard errors to weight the observations
         for i in range(ntimes):
             
             # Only (re-)calculate SVD when necessary
-            if i == 0 or not np.all(obs_var[i] == obs_var[i-1]):
+            if i == 0 or not np.all(obs_std[i] == obs_std[i-1]):
 
-                # Weight T_obs with obs_var
+                # Weight T_obs with obs_std
                 svd_in = (T_obs.reshape(-1, self.nsec) /
-                          obs_var[i].ravel()[:, np.newaxis])
+                          obs_std[i].ravel()[:, np.newaxis])
 
                 # Find singular value decompostion
                 U, S, Vh = np.linalg.svd(svd_in, full_matrices=False)
 
-                # Divide by infinity (1/S) gives zero weights
+                # Eliminate singular values less than epsilon by setting their
+                # reciprocal to zero (setting S to infinity firsts avoids
+                #  divide-by-zero warings)
+                S[S < epsilon * S.max()] = np.inf
                 W = 1./S
 
-                # Eliminate the small singular values (less than epsilon)
-                # by giving them zero weight
-                W[S < epsilon*S.max()] = 0.
-
-                # Update VWU if obs_var changed
+                # Update VWU if obs_std changed
                 VWU = Vh.T @ (np.diag(W) @ U.T)
 
             # Solve for SEC amplitudes and error variances
             # shape: ntimes x nsec
-            self.sec_amps[i, :] = (VWU @ (obs_B[i]/obs_var[i]).reshape(-1).T).T
+            self.sec_amps[i, :] = (VWU @ (obs_B[i] / obs_std[i]).reshape(-1).T).T
 
             # Maybe we want the variance of the predictions sometime later...?
             # shape: ntimes x nsec
-            self.sec_amps_var[i, :] = np.sum((Vh.T * W)**2, axis=1)
+            valid = np.isfinite(obs_std[i].reshape(-1))
+            self.sec_amps_var[i, :] = np.sum(
+                (VWU[:,valid] * obs_std[i].reshape(-1)[valid])**2,
```

Comment:
Do we want to store the amps variance, or should we change this to standard error as well?

I will need to wrap my head around why we are changing over to `VWU` here now instead of `Vh.T * W`. At a quick glance it looks like what we were doing before is taking the variance from the SVD solution (effectively the error associated with geometric transforms?) and mapping that to each SEC location. Now, you are weighting the SVD transform by the standard error associated with that geometric transform, so you're actually propagating the observation standard errors geometrically through the system now?

Reply:
I think you should keep secs.sec_amps_var, and NOT change it to secs.sec_amps_std. My reasoning is that, typically, when one propagates uncertainty through a linear operator, it is presented in terms of a (co)variance matrix of the data being operated on (see: https://en.wikipedia.org/wiki/Propagation_of_uncertainty#linear_combinations). In our case, it's trivial to sqrt the diagonal matrix, but if we ever choose to use the full covariance, things get more complicated. 

In spite of that logic, changing the input to obs_std still seemed like a good idea because I think most scientists have an intuitive sense of "weighted least-squares" based on the (independent) standard error of observations, but frankly, wouldn't know what to do with an actual "covariance matrix". I might be underestimating many of our peers.

To answer your other question, the "variance from the SVD solution" would be the square of the singular values (the S's in secs.fit(), or 1/W), correct? This is still there, just wrapped up in **VWU**. That's all we had previously, and it is what was wrong (or incomplete) in Numerical Recipes. It was necessary to add the observation error (co)variance to complete things, which is what the Mandel-1982 paper pointed me toward, even if it didn't do a great job explaining things.

Finally, if you visit https://math.stackexchange.com/a/1169843, toward the bottom you will see why **VWU** should be used instead of the summation of **Vh**<sup>T</sup> * **W**. The latter is only valid if all observation error variances are IID and unitary. That is, the observation error covariance matrix is diagonal, and all the entries are 1.

It's a bit confusing because the author of that StackExchange answer labels his covariance matrix **I**, and generally treats it like it actually is an identity matrix, which he justifies by saying if it isn't, you can make it so by "scaling" the equations, which just means dividing each observation by its standard deviation.

But if you keep the whole expression **V** @ **W** @ **U**<sup>T</sup> @ **I** @ **U** @ **W** @ **V**<sup>T</sup> (where **W** = **S**<sup>-1</sup>), you get what you would expect from applying standard error propagation through a linear operation (again, see: https://en.wikipedia.org/wiki/Propagation_of_uncertainty#linear_combinations).

Since we already calculated **VWU** = **V** @ **W** @ **U**<sup>T</sup>, and because **I** is NOT identity, we square and multiply **VWU** * **I** element-wise, then sum across the columns to efficiently get the diagonal elements of the covariance matrix. If we wanted the full covariance matrix (which would be ideal, but prohibitive both memory-wise and cpu-wise), we would do **[VWU]** @ **I** @ **[VWU]**<sup>T</sup>.

Reply:
Wow, that is an excellent write-up. I think we should try and capture all of that information somewhere in the code. Maybe in documentation somewhere?